src/signal.cr

```diff
@@ -90,67 +25,215 @@ require "c/unistd"
-    Event::SignalHandler.del_handler self
-    LibC.signal value, block
-  end
-
-  @@default_handlers_setup = false
+  @@setup_default_handlers = Atomic(Int32).new(0)
 
   # :nodoc:
   def self.setup_default_handlers
-    return if @@default_handlers_setup
-    @@default_handlers_setup = true
+    _, success = @@setup_default_handlers.compare_and_set(0, 1)
+    return unless success
 
+    Crystal::SignalHandler.spawn
     Signal::PIPE.ignore
     Signal::CHLD.reset
   end
 end
 
 # :nodoc:
-fun __crystal_sigfault_handler(sig : LibC::Int, addr : Void*)
-  Crystal.restore_blocking_state
+module Crystal::SignalHandler
+  # The number of libc functions that can be called safely from a signal(2)
+  # handler is very limited. An usual safe solution is to use a pipe(2) and
+  # just write the signal to the file descriptor and nothing more. A loop in
+  # the main program is responsible for reading the signals back from the
+  # pipe(2) and handle the signal there.
+
+  @@pipe = IO.pipe(read_blocking: false, write_blocking: true)
+  @@procs = {} of Signal => Signal ->
+  @@mutex = Mutex.new
+
+  def self.trap(signal, handler) : Nil
+    @@mutex.synchronize do
+      unless @@procs[signal]?
+        LibC.signal(signal.value, ->(value : Int32) {
+          writer.write_bytes(value)
+        })
+      end
+      @@procs[signal] = handler
+    end
+  end
+
+  def self.reset(signal) : Nil
+    set(signal, Proc(Int32, Void).new(Pointer(Void).new(0_u64), Pointer(Void).null))
+  end
 
-  # Capture fault signals (SEGV, BUS) and finish the process printing a backtrace first
-  LibC.dprintf 2, "Invalid memory access (signal %d) at address 0x%lx\n", sig, addr
-  CallStack.print_backtrace
-  LibC._exit sig
+  def self.ignore(signal) : Nil
+    set(signal, Proc(Int32, Void).new(Pointer(Void).new(1_u64), Pointer(Void).null))
+  end
+
+  private def self.set(signal, handler)
+    if signal == Signal::CHLD
+      # don't reset/ignore SIGCHLD, Process#wait requires it
+      trap(signal, ->(signal : Signal) {
+        Crystal::SignalChildHandler.call
+      })
+    else
+      @@mutex.synchronize do
+        @@procs.delete(signal)
+        LibC.signal(signal.value, handler)
+      end
+    end
+  end
+
+  def self.spawn
+    ::spawn do
+      loop do
+        value = reader.read_bytes(Int32)
+        process(Signal.new(value))
+      end
+    end
+  end
+
+  def self.process(signal) : Nil
+    if handler = @@procs[signal]?
+      handler.call(signal)
+    else
+      fatal("missing handler for #{signal}")
+    end
+  rescue ex
+    ex.inspect_with_backtrace(STDERR)
+    fatal("uncaught exception while processing handler for #{signal}")
+  end
+
+  def self.after_fork
+    @@pipe = IO.pipe(read_blocking: false, write_blocking: true)
+  end
+
+  private def self.reader
+    @@pipe[0]
+  end
+
+  private def self.writer
+    @@pipe[1]
+  end
+
+  private def self.fatal(message : String)
+    STDERR.puts("FATAL: #{message}, exiting")
+    STDERR.flush
+    LibC._exit(1)
```

Comment:
Why not normal exit? We're not in a signal handler right now.

Reply:
I kept this behavior, but I have mixed feelings about it:

1. this is supposed to be a "fatal error" so it's probably right to exit immediately, skipping atexit handlers;
2. maybe an uncaught exception in a signal handler shouldn't be fatal; it's just an uncaught exception in a regular fiber (we're not in `altsigstack`), but if a SIGINT handler raises, there will be no way to kill the process without SIGKILL for example.

I'm not sure what's correct, here. Spat an error to STDOUT and continue on? Exit immediately? Exit and run atexit handlers? All of them can be somewhat valid (and more importantly safe).

Reply:
The problem I have is the proliferation of calls to LibC.exit functions, when the runtime imposes many things we need to do at exit. We always should try to restore blocking state. We might want to run at exit handlers. But either way there should be a single exit point for every application, so that we can actually debug what's happening when the application exits instead of have to copy things like restoring the blocking state to every exit point.

Reply:
Good point. Maybe a global `exit` immediately (and `abort(message)`) with some limited but required system cleanup could be nice? Sometimes skipping atexit handlers is useful (and sending SIGABRT).

What should be done here? Just `exit(1)`?

Reply:
For now I would keep this as-is and whack in a call to restore blocking state. Then make a issue and pr to solve this for real.